File: backend/src/http.ts
import cors from "cors";
import express from "express";
import { ZodError } from "zod";

import { analyticsRoutes } from "./controllers/analytics.js";
import { AnalyticsService } from "./services/analytics.js";
import { employeeRoutes } from "./controllers/employees.js";
import type { EmployeeRepository } from "./database/repos/employees.js";
import { EmployeeService } from "./services/employees.js";
import type { FxRepository } from "./database/repos/fx.js";
import { FxService } from "./services/fx.js";
import type { SalaryRepository } from "./database/repos/salaries.js";
import { SalaryService } from "./services/salaries.js";
import {
  COUNTRY_CODES,
  DEPARTMENTS,
  STATUSES,
  SUPPORTED_CURRENCIES,
} from "./common/constants.js";
import { calendarDateToday } from "./common/dates.js";
import { ConflictError, DomainError, NotFoundError } from "./common/errors.js";

export function createApp(deps: {
  employees: EmployeeRepository;
  salaries: SalaryRepository;
  rates: FxRepository;
  today?: () => string;
}) {
  const today = deps.today ?? calendarDateToday;
  const employees = new EmployeeService(deps.employees);
  const salaries = new SalaryService(deps.employees, deps.salaries);
  const fx = new FxService(deps.rates);
  const analytics = new AnalyticsService(deps.employees, deps.salaries, fx);

  const corsOrigin = process.env.VERCEL
    ? true
    : ["http://localhost:5173", "http://127.0.0.1:5173"];

  const app = express();
  app.use(cors({ origin: corsOrigin }));
  app.use(express.json({ limit: "1mb" }));

  app.get("/api/health", (_req, res) => {
    res.json({ status: "ok" });
  });
  app.get("/api/meta", (_req, res) => {
    res.json({
      country_codes: [...COUNTRY_CODES],
      departments: [...DEPARTMENTS],
      currencies: [...SUPPORTED_CURRENCIES],
      statuses: [...STATUSES],
    });
  });
  app.use(
    "/api/employees",
    employeeRoutes({ employees, salaries, today }),
  );
  app.use("/api/analytics", analyticsRoutes({ analytics, today }));

  app.use((error: unknown, _req: express.Request, res: express.Response, _next: express.NextFunction) => {
    if (error instanceof ZodError) {
      res.status(422).json({ detail: error.issues[0]?.message ?? "invalid input", code: "invalid" });
      return;
    }
    if (error instanceof NotFoundError) {
      res.status(404).json({ detail: error.message, code: error.code });
      return;
    }
    if (error instanceof ConflictError) {
      res.status(409).json({ detail: error.message, code: error.code });
      return;
    }
    if (error instanceof DomainError) {
      res.status(400).json({ detail: error.message, code: error.code });
      return;
    }
    console.error(error);
    res.status(500).json({ detail: "internal error", code: "internal" });
  });

  return app;
}

File: backend-py/app/http.py
from __future__ import annotations

from collections.abc import Callable

from fastapi import FastAPI, Request
from fastapi.exceptions import RequestValidationError
from fastapi.middleware.cors import CORSMiddleware
from fastapi.responses import JSONResponse

from app.constants import COUNTRY_CODES, DEPARTMENTS, STATUSES, SUPPORTED_CURRENCIES
from app.dates import calendar_date_today
from app.errors import ConflictError, DomainError, NotFoundError
from app.repos import EmployeeRepository, FxRepository, SalaryRepository
from app.routers.analytics import router as analytics_router
from app.routers.employees import router as employees_router
from app.services.analytics import AnalyticsService
from app.services.employees import EmployeeService
from app.services.fx import FxService
from app.services.salaries import SalaryService


def create_app(
    employees: EmployeeRepository,
    salaries: SalaryRepository,
    rates: FxRepository,
    today: Callable[[], str] | None = None,
) -> FastAPI:
    today_fn = today or calendar_date_today
    employee_service = EmployeeService(employees)
    salary_service = SalaryService(employees, salaries)
    fx = FxService(rates)
    analytics = AnalyticsService(employees, salaries, fx)

    app = FastAPI(title="ACME Pay API")
    app.add_middleware(
        CORSMiddleware,
        allow_origins=[
            "http://localhost:5173",
            "http://127.0.0.1:5173",
            "https://acme-pay.vercel.app",
        ],
        allow_origin_regex=r"https://.*\.vercel\.app",
        allow_credentials=True,
        allow_methods=["*"],
        allow_headers=["*"],
    )
    app.state.employees = employee_service
    app.state.salaries = salary_service
    app.state.analytics = analytics
    app.state.today = today_fn

    @app.exception_handler(RequestValidationError)
    async def validation_handler(_request: Request, exc: RequestValidationError):
        first = exc.errors()[0] if exc.errors() else {}
        return JSONResponse(
            {"detail": first.get("msg", "invalid input"), "code": "invalid"},
            status_code=422,
        )

    @app.exception_handler(NotFoundError)
    async def not_found_handler(_request: Request, exc: NotFoundError):
        return JSONResponse({"detail": str(exc), "code": exc.code}, status_code=404)

    @app.exception_handler(ConflictError)
    async def conflict_handler(_request: Request, exc: ConflictError):
        return JSONResponse({"detail": str(exc), "code": exc.code}, status_code=409)

    @app.exception_handler(DomainError)
    async def domain_handler(_request: Request, exc: DomainError):
        return JSONResponse({"detail": str(exc), "code": exc.code}, status_code=400)

    @app.get("/api/health")
    def health():
        return {"status": "ok"}

    @app.get("/api/meta")
    def meta():
        return {
            "country_codes": list(COUNTRY_CODES),
            "departments": list(DEPARTMENTS),
            "currencies": list(SUPPORTED_CURRENCIES),
            "statuses": list(STATUSES),
        }

    app.include_router(employees_router)
    app.include_router(analytics_router)
    return app
